### PR TITLE
Speedy: refactor Machine builders

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR we make clean-up the constructor for the speedy Machine.
    
* We remove the `case`  keyword since `Machine` is a stateful class,
* We replace the pre-existing builders with
      + one generic builder `Machine.apply`,
      + simple scenario specific builder,
      + simple non-scenario specific builder,
    

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -546,73 +551,29 @@ object Speedy {
...
-        sexpr: SExpr,
+        seed: crypto.Hash,
+        expr: SExpr,
+    ): Machine = Machine(
+      compiledPackages,
+      Time.Timestamp.MinValue,
+      InitialSeeding.TransactionSeed(seed),
+      SEApp(expr, Array(SEValue.Token)),
+      Set.empty,
+      Set.empty,
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:564]: Maybe named arguments make this easier to read?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/InterpreterTest.scala [hunk 2]
```diff
@@ -22,13 +24,16 @@ class InterpreterTest extends WordSpec with Matchers with TableDrivenPropertyChe
...
-      expr = e,
-      compiledPackages = PureCompiledPackages(Map.empty).right.get,
-      scenario = false,
+    val machine = Speedy.Machine(
+      compiledPackages = noPackages,
       submissionTime = Time.Timestamp.now(),
       initialSeeding = InitialSeeding.NoSeed,
+      noPackages.compiler.unsafeCompile(e),
+      Set.empty,
+      Set.empty
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/InterpreterTest.scala:36]: Named arguments for the last three args as well?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/InterpreterTest.scala [hunk 3]
```diff
@@ -127,21 +132,21 @@ class InterpreterTest extends WordSpec with Matchers with TableDrivenPropertyChe
...
-        scenario = false,
-        submissionTime = Time.Timestamp.now(),
-        initialSeeding = InitialSeeding.NoSeed,
+      machine = Speedy.Machine(
+        noPackages,
+        Time.Timestamp.now(),
+        InitialSeeding.NoSeed,
+        list,
+        Set.empty,
+        Set.empty
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/InterpreterTest.scala:149]: Named arguments?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala [hunk 4]
```diff
@@ -305,14 +305,18 @@ class SpeedyTest extends WordSpec with Matchers {
...
-      expr = e,
+  private[this] def build(e: Expr, packages: PureCompiledPackages) =
+    Speedy.Machine(
       compiledPackages = packages,
-      scenario = false,
       submissionTime = Time.Timestamp.now(),
       initialSeeding = InitialSeeding.NoSeed,
+      expr = packages.compiler.unsafeCompile(e),
+      globalCids = Set.empty,
+      committers = Set.empty
```
Comment (hurryabit) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala:315]: Can't we use `build` here (like below) as well?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/InterpreterTest.scala (hunk 2 above)
Comment (hurryabit) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/InterpreterTest.scala:36]: Should we have a convenience function akin to `Machine.buildForScenario` for the pattern with `Timestamp.now()`,`NoSeed` and two times `Set.empty` as well? This appears a few times.

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala [hunk 5]
```diff
@@ -419,13 +418,15 @@ object Repl {
...
             val machine =
-              Speedy.Machine.fromExpr(
-                expr = expr,
-                compiledPackages = PureCompiledPackages(state.packages).right.get,
-                scenario = false,
+              Speedy.Machine(
+                compiledPackages = compiledPackages,
                 submissionTime = Time.Timestamp.now(),
                 initialSeeding = InitialSeeding.NoSeed,
+                expr = SEApp(compiledPackages.compiler.unsafeCompile(expr), Array(SEValue.Token)),
```
Comment (cocreature) [inline daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:427]: Why do we have a token here now? That doesn’t look right.

Comment (remyhaemmerle-da) [reply daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:427]: Good catch

**Pull request merged**